org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/NullAnnotationTests18.java

```diff
@@ -541,4 +541,54 @@ public void testBug576329() {
 		runner.expectedOutputString = "0";
 		runner.runConformTest();
 	}
+
+	@org.junit.Test
+	public void testNamedInstanceOfIsNonNull() {
+		Runner runner = getDefaultRunner();
+		runner.testFiles = new String[] {
+				"X.java",
+				"import org.eclipse.jdt.annotation.*;\n" +
+				"public class X {\n" +
+				"	public static void consumeNonNull(@NonNull String s) {\n" +
+				"		System.out.println(\"nonnull\");\n" +
+				"	}\n" +
+				"	public static void main(String... args) {\n" +
+				"		Object o = Math.random() < 0 ? new Object() : \"blah\";\n" +
+				"		if (o instanceof String message) {\n" +
+				"			consumeNonNull(message);\n" +
+				"		}\n" +
+				"	}\n" +
+				"}\n"
+			};
+		runner.expectedCompilerLog = "";
+		runner.expectedOutputString = "nonnull";
+		runner.runConformTest();
+		//
```

Comment:
nitpick: we normally prefer one test per test method, unless there's a very good excuse like reusing lots of test data.